src/FSharpPlus/Data/Matrix.fs

```diff
@@ -0,0 +1,643 @@
+  /// matrix product
+  static member inline ( .* ) (m1, m2) = Matrix.matrixProduct m1 m2
+  /// kronecker (tensor) product
+  static member inline ( @* ) (m1, m2) = Matrix.kroneckerProduct m1 m2
+  /// hadamard (element-wise) product
+  static member inline ( * ) (m1, m2) = Matrix.map2 (*) m1 m2
+  static member inline ( / ) (m1, m2) = Matrix.map2 (/) m1 m2
+  static member inline ( * ) (m: Matrix<'a,_,_>, s: 'a) = Matrix.map ((*) s) m
+  static member inline ( * ) (s: 'a, m: Matrix<'a,_,_>) = Matrix.map ((*) s) m
+  static member inline ( / ) (m: Matrix<'a,_,_>, s: 'a) = Matrix.map (fun x -> x / s) m
+  static member inline ( / ) (s: 'a, m: Matrix<'a,_,_>) = Matrix.map (fun x -> s / x) m
+  static member inline ( @| ) (m1: Matrix<_,'m,^n1>, m2: Matrix<_,'m,^n2>) : Matrix<_,'m,^``n1 + ^n2``> = Matrix.horizontalSum m1 m2
+  static member inline ( @- ) (m1: Matrix<_,^m1,'n>, m2: Matrix<_,^m2,'n>) : Matrix<_,^``m1 + ^m2``,'n> = Matrix.verticalSum m1 m2
+  static member inline ( ~- ) m = Matrix.map ((~-)) m
+
+  // As generic number literals
+  static member inline FromInt32  (i: int   ) = Matrix.replicate Singleton Singleton (FromInt32.Invoke  i)
+  static member inline FromInt64  (i: int64 ) = Matrix.replicate Singleton Singleton (FromInt64.Invoke  i)
+  static member inline FromBigInt (i: bigint) = Matrix.replicate Singleton Singleton (FromBigInt.Invoke i)
+
+
+type Vector<'Item, 'Length> with
+  static member inline Item (v: Vector<'a, 'n>, i) = Vector.get i v
+  static member inline Map (v: Vector<'a, 'n>, f: 'a -> 'b) : Vector<'b, 'n> = Vector.map f v
+  static member inline Return (x: 'x) : Vector<'x, 'n> = Vector.replicate Singleton x
+  static member inline ( <*> ) (f: Vector<'x -> 'y, 'n>, x: Vector<'x, 'n>) : Vector<'y, 'n> = Vector.apply f x
+  static member inline Zip (x, y) = Vector.zip x y
+  static member inline get_Zero () : Vector<'x, 'n> = Vector.zero
+  static member inline ( + ) (v1: Vector<_, 'n>, v2: Vector<_, 'n>) = Vector.map2 (+) v1 v2
+  static member inline ( - ) (v1: Vector<_, 'n>, v2: Vector<_, 'n>) = Vector.map2 (-) v1 v2
+  /// dot (inner) product
+  static member inline ( .* ) (v1, v2) = Vector.innerProduct v1 v2
+  /// direct (tensor) product
+  static member inline ( @* ) (v1, v2) = Vector.directProduct v1 v2
+  /// cross product
+  static member inline ( %* ) (v1, v2) = Vector.vectorProduct3 v1 v2
+  /// hadamard (element-wise) product
+  static member inline ( * ) (v1: Vector<_, 'n>, v2: Vector<_, 'n>) = Vector.map2 (*) v1 v2
+  static member inline ( / ) (v1: Vector<_, 'n>, v2: Vector<_, 'n>) = Vector.map2 (/) v1 v2
+  static member inline ( * ) (v: Vector<'a, 'n>, s: 'a) = Vector.map (fun x -> x * s) v
+  static member inline ( * ) (s: 'a, v: Vector<'a, 'n>) = Vector.map (fun x -> x * s) v
+  static member inline ( / ) (v: Vector<'a, 'n>, s: 'a) = Vector.map (fun x -> x / s) v
+  static member inline ( / ) (s: 'a, v: Vector<'a, 'n>) = Vector.map (fun x -> s / x) v
```

Comment:
Here I see problem.
It's my understanding that we shouldn't add overloads for binary operations accepting a scalar and a vector, by doing so we're complicating type inference to a point where it's not decidable.
In order to multiply (or divide) by a scalar there are many other ways, the easiest one being to use generic literals (see the example I added in the docs).

Reply:
OK, we can use `result` as in the doc anyway.

Reply:
Yes, `return`, `map` / `<!>`, applicative math operators or generic literals. But not automatically.
I think generic literals is the easiest, it requires only one additional keystroke.